WooCommerce/Classes/ViewModels/Order Details/OrderDetailsDataSource.swift

```diff
@@ -24,10 +24,18 @@ final class OrderDetailsDataSource: NSObject {
     ///
     private(set) var sections = [Section]()
 
-    /// Is this order processing?
+    /// Does this order need processing? Payment received (paid). The order is awaiting fulfillment.
+    /// All product orders require processing, except those that only contain products which are both Virtual and Downloadable.
     ///
-    private var isProcessingPayment: Bool {
-        return order.status == OrderStatusEnum.processing
+    private var needsProcessing: Bool {
```

Comment:
I have a few nonblocker suggestions for this computed property:
- Since the body of the property cannot be executed in O(1), I would use a method for this one. That shows some non trivial computation inside, implying that it should be used with care to avoid performance issues. More [here](https://www.swiftbysundell.com/tips/computed-properties-vs-methods/)
- I would remove `self` when possible
- We can use `guard` for `order.needsProcessing`
- We can skip using the `return` keyword when we have only one expression in our function
- We can break the long `orderContainsOnlyVirtualAndDownloadableProducts` into shorter ones to make it more readable
- We can use `contains`  instead of `first` and `!= nil` to make it shorter
- I would extract the `allOrderProductsAreVirtualAndDownloadable` logic into its own function to make more readable

With all these suggestions we can have something like 

```
 private func needsProcessing() -> Bool {
        guard !order.needsProcessing else {
            return true
        }

        return !allOrderProductsAreVirtualAndDownloadable()
 }

private func allOrderProductsAreVirtualAndDownloadable() -> Bool {
        let orderProducts = products.filter { (product) -> Bool in
            items.contains(where: { $0.productID == product.productID })
        }

        return orderProducts.allSatisfy { $0.virtual && $0.downloadable }
    }
```

Reply:
This is great, thanks for all the details and the link. Changed here: 0362f6f3a6fa2dd6f2cc1223cc9188d9535cfbd7